src/sssom/util.py

```diff
@@ -165,7 +165,15 @@ def from_mapping_set_document(cls, doc: MappingSetDocument) -> "MappingSetDataFr
             # For pandas < 2.0.0, call 'infer_objects()' without any parameters
             df = df.infer_objects()
         # remove columns where all values are blank.
+
+        # Context https://github.com/pandas-dev/pandas/issues/57734
+        try:
+            pd.set_option("future.no_silent_downcasting", True)
+        except KeyError:
+            # Option does not exist in this version of pandas
+            pass
         df.replace("", np.nan, inplace=True)
+        # df.infer_objects(copy=False)
```

Comment:
`infer_objects` does not work “in place” (and it doesn’t have a `inplace` keyword to force it to do so), so that call does nothing since you’re not collecting the results.

You’d need

```python
df = df.infer_objects(copy=False)
```

As for compatibility with various versions of Pandas: the `copy` parameter was introduced in Pandas 2.0. You could leave it out:

```python
df = df.infer_objects()
```

Without that `copy=False`, the default behaviour of `infer_objects` would be to make a copy of all values in the frame, _even if they are not casted to another type_. But this is not a concern here since you are not keeping the original frame anyway.

As for whether the downcasting is necessary to begin with, well, that would depend on the rest of the code downstream from here. For now (before this PR), the data frame _is_ downcasted, as a side-effect of `replace`. The question is, is there some code downstream that is depending on the dataframe columns having more precise types than `object`?

Reply:
I added it back with a pandas version check, can you see what you think?

Reply:
This is fine, but as I said you could also have just left the `copy` parameter out.

For users of Pandas 1.x, this does not change anything.

For users of Pandas 3.x, this also does not change anything, since copy-on-write means that even if a copy of the non-type-casted objects is made, the actual copy would never actually take place (since we’re re-assigning the inferred data frame to the original variable).

For users of Pandas 2.x, no `copy=False` parameter means that the non-type-casted objects would be copied for nothing, which strictly speaking is a waste of time and memory. But I strongly suspect the effects on performances would be negligible.

Reply:
(Side note: I _love_ the “should GitHub still exist when you read this”. :D  )

Reply:
For now I just want to do the thing that changes least about the existing behaviour - which I thought I understand form your explanations would be `copy=False` for Pandas 2.x?

Reply:
If you do

```python
df = df.infer_objects(copy=False)
```

then `copy=False` makes absolutely no difference to the _result_. All it does it makes the operation slightly faster (which is why I included it in my example on Slack, but I now regret having done that if it led you to believe it was absolutely necessary), but the data frame you get at the end will be exactly the same as if you omitted the `copy` parameter.

What does the `copy` parameter does exactly?

Consider the following data frame:

```python
>>> df
   A         B
0  alice     1
1  bob       2
2  charlie     
```

Let’s create `df1` where we replace the empty cell (B2) with `np.nan`:

```python
>>> pd.set_option('future.no_silent_downcasting', True)
>>> df1 = df.replace("", np.nan)
>>> df1
   A         B
0  alice     1
1  bob       2
2  charlie   np.nan
```

Now let’s downcast, first with `copy=True` (default behaviour, same as no `copy` parameter at all):

```python
>>> df2_copy_true = df1.infer_objects(copy=True)
>>> df2_copy_true
   A         B
0  alice     1.0
1  bob       2.0
3  charlie   np.nan
```

Then with `copy=False`:


```python
>>> df2_copy_false = df1.infer_objects(copy=True)
>>> df2_copy_false
   A         B
0  alice     1.0
1  bob       2.0
3  charlie   np.nan
```

Observe that `df2_copy_true` and `df2_copy_false` are the same. So what’s the difference?

The difference is that the *A* column in `df2_copy_false` share its data with the _A_ column in the original `df1` frame: A copy was not made, since there was no downcasting needed for that column.

This means that if you modify in place a value in the *A* column of `df2_copy_false`, this will _also_ modify the original `df1`:

```python
>>> df2_copy_false.replace("alice", "arthur", inplace=True)
>>> df2_copy_false
   A         B
0  arthur    1.0
1  bob       2.0
3  charlie   np.nan
>>> df1
   A         B
0  arthur    1
1  bob       2
2  charlie   np.nan
```

By contrast, `df2_copy_true` is completely distinct from the original `df1`. Even if the *A* column needed no downcast, `infer_objects(copy=True)` made a complete copy of it.

From the above, you should understand that you do not need to worry about getting different results depending on the presence of `copy=False` if all you do is re-assigning the output of `infer_objects()` to the original variable – after the assignment the original data frame is no longer there anyway.

So when you do this:

```python
df = df.infer_objects()
```

The *only* impact of the absence of `copy=False` is that you will needlessly make a temporary copy of columns that didn’t need to be copied, so it’s _slightly_ less efficient than

```python
df = df.infer_objects(copy=False)
```

Whether you should be concerned about this inefficiency is up to you. I am personally not concerned enough to think that it warrants making the code more complicated by adding this `_safe_infer_objects` intermediate, but it’s your call.

In any case, the inefficiency will go away with Pandas 3.0 and its copy-on-write mechanism.

Reply:
> In any case, the inefficiency will go away with Pandas 3.0 and its copy-on-write mechanism.

I would in fact be keen to enable copy-on-write globally (`pd.set_option('mode.copy_on_write', True)`), both for the performance improvement and to start making sure that we are ready to deal with it since it will become mandatory with Pandas 3.0.

But the concerns [stated above](https://github.com/mapping-commons/sssom-py/pull/561#discussion_r1835342738) for the people who use `sssom` as a Python library would obviously also apply here.

Reply:
> I would in fact be keen to enable copy-on-write globally

Wait, we are in fact _already_ [enabling copy-on-write](https://github.com/mapping-commons/sssom-py/blob/4cad7d6d8b905728f14a90d3d4c6d591b520cd3b/src/sssom/util.py#L86):

```python
pd.options.mode.copy_on_write = True
```

So there’s definitely no need for `copy=False` when we call `infer_objects()` here.

Reply:
Thanks so much for taking the time to explaining this! Really much appreciated!